### PR TITLE
adding autofill gas option sendtransaction

### LINKED ISSUE #6215: maxFeePerGas requirement in 4.x
### Discussed in https://github.com/web3/web3.js/discussions/6188

<div type='discussions-op-text'>

<sup>Originally posted by **t-vila** June 12, 2023</sup>
Hey all,

As far as I remember `maxFeePerGas` was optional in 1.x and calculated from the `baseFee` - in 4.x I get `MissingGasError` if I don't specify it. It looks like it's expected https://github.com/web3/web3.js/blob/f8a2533c2b09ce0a62f8414f2f6eed83ab78ca1f/packages/web3-errors/src/errors/transaction_errors.ts#L315

For instance this:

```
const tx = {
    from: signer.address,
    to: "...",
    value: web3.utils.toWei('0.0001', 'ether'),
    gas: limit,
    nonce: await web3.eth.getTransactionCount(signer.address),
    maxPriorityFeePerGas: web3.utils.toWei('3', 'gwei'),
    //maxFeePerGas: web3.utils.toWei('3', 'gwei'),
    chainId: 11155111,
    type: 0x2
  };
```
Will throw:

`MissingGasError: Invalid value given "gas: 0x5208, gasPrice: undefined, maxPriorityFeePerGas: 0xb2d05e00, maxFeePerGas: undefined". Error: "gas" is missing.`</div>

### LINKED ISSUE #6249: Migration from 1.8v to 4v caused Insufficient funds error
I have upgraded web3 and web3-eth from 1.8.0 to 4.0.2 (mostly because of less package size)
I've read the migration guide and, since I had quite a simple logic of connecting wallet + sending transaction, there was not much to change

I haven't changed anything in web3.eth.sendTransaction, but after updating I started getting this error for every single transaction (even though there are funds):

```
Returned error: err: insufficient funds for gas * price + value: address 0x568A33aA237fFc284FF584A3BAB17A44FAeF4054 have 61046238310789005 want 8879502456500000000 (supplied gas 500000000)
```

Here is the code for sendTransaction:
```
return web3.eth.sendTransaction({
						from: selectedAccount,
						to: window.toonAppData.address,
						data: "0x64edfbf0e2c706ba4a09595315c45355a341a576cc17f3a19f43ac1c02f814ee",
						value: web3.utils.toWei(amount, "ether"),
					})
```

I'm quite new to web3 and I'm not the one who wrote the initial code, but since I had to update wallet connect, this issue also appeared...

#### Environment
"devDependencies": {
		"@babel/core": "^7.20.2",
		"@babel/preset-env": "^7.20.2",
		"@babel/preset-react": "^7.22.5",
		"@types/react": "^18.2.14",
		"@types/react-dom": "^18.2.6",
		"assert": "^2.0.0",
		"autoprefixer": "^10.4.13",
		"babel-loader": "^9.1.0",
		"babel-preset-es2015": "^6.24.1",
		"buffer": "^6.0.3",
		"crypto-browserify": "^3.12.0",
		"css-loader": "^6.7.1",
		"css-minimizer-webpack-plugin": "^4.2.2",
		"fs-extra": "^10.1.0",
		"html-minimizer-webpack-plugin": "^4.3.0",
		"html-webpack-plugin": "^5.5.0",
		"https-browserify": "^1.0.0",
		"less": "^4.1.3",
		"less-loader": "^11.1.0",
		"mini-css-extract-plugin": "^2.6.1",
		"os-browserify": "^0.3.0",
		"postcss-loader": "^7.0.1",
		"process": "^0.11.10",
		"sass": "^1.60.0",
		"sass-loader": "^13.2.2",
		"stream-browserify": "^3.0.0",
		"stream-http": "^3.2.0",
		"style-loader": "^3.3.1",
		"url": "^0.11.0",
		"webpack": "^5.74.0",
		"webpack-cli": "^4.10.0",
		"webpack-dev-server": "^4.11.1"
	},
	"dependencies": {
		"@metamask/detect-provider": "^2.0.0",
		"@wagmi/core": "^1.3.3",
		"@walletconnect/ethereum-provider": "^2.8.6",
		"@web3modal/ethereum": "^2.6.2",
		"@web3modal/html": "^2.6.2",
		"browserify-zlib": "^0.2.0",
		"path-browserify": "^1.0.1",
		"react": "^18.2.0",
		"react-dom": "^18.2.0",
		"react-error-boundary": "^4.0.10",
		"viem": "^1.2.6",
		"web3": "^4.0.2",
		"web3-eth": "^4.0.2"
	},

### PR DESCRIPTION
**Pull request by luu-alex**
## Description

Please include a summary of the changes and be sure to follow our [Contribution Guidelines](../CONTRIBUTIONS.md).

Fixes #6215, updated error message to be more clear what gas property needs to be defined
Fixes #6249, updates transaction builder to automatically fill gas in the `sendTransaction` method

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] I have selected the correct base branch.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran `npm run lint` with success and extended the tests and types if necessary.
- [ ] I ran `npm run test:unit` with success.
- [ ] I ran `npm run test:coverage` and my test cases cover all the lines and branches of the added code.
- [ ] I ran `npm run build` and tested `dist/web3.min.js` in a browser.
- [ ] I have tested my code on the live network.
- [ ] I have checked the Deploy Preview and it looks correct.
- [ ] I have updated the `CHANGELOG.md` file in the root folder.
- [ ] I have linked Issue(s) with this PR in "Linked Issues" menu.

### CODE REVIEW

**Review by avkos — CHANGES_REQUESTED** [review]
LGTM, but tests fail

File: packages/web3-errors/src/errors/transaction_errors.ts [hunk 1]
```diff
@@ -327,7 +327,7 @@ export class MissingGasError extends InvalidValueError {
 			}, maxPriorityFeePerGas: ${value.maxPriorityFeePerGas ?? 'undefined'}, maxFeePerGas: ${
 				value.maxFeePerGas ?? 'undefined'
 			}`,
-			'"gas" is missing',
+			'Missing properties in transaction, either define "gas" and "gasPrice" for type 0 transactions or "gas", "maxPriorityFeePerGas" and "maxFeePerGas" for type 2 transactions',
```
Comment (Spacesai1or) [inline packages/web3-errors/src/errors/transaction_errors.ts]: Not suggesting we revert this change, but this is technically a breaking change since `'"gas" is missing'` was the error message thrown in `1.x` (and current `4.x`)

**Review by Spacesai1or — COMMENTED** [review]
LGTM besides previous comment and failing tests

File: packages/web3-errors/src/errors/transaction_errors.ts [hunk 2]
```diff
@@ -347,7 +347,7 @@ export class TransactionGasMismatchError extends InvalidValueError {
 			}, maxPriorityFeePerGas: ${value.maxPriorityFeePerGas ?? 'undefined'}, maxFeePerGas: ${
 				value.maxFeePerGas ?? 'undefined'
 			}`,
-			'transaction must specify legacy or fee market gas properties, not both',
+			'Transaction must define "gas" and "gasPrice" for type 0 transactions or "gas", "maxPriorityFeePerGas" and "maxFeePerGas" for type 2 transactions, not both',
```
Comment (jdevcs) [inline packages/web3-errors/src/errors/transaction_errors.ts]: Along with last one, this will also be breaking change as compared to current 4.x

Comment (jdevcs) [reply packages/web3-errors/src/errors/transaction_errors.ts]: Can we add some inner error ( for more details ) ? so we dnt break API, and add more details in that inner error. or do you want to propose any other way of doing it without breaking changes?

File: packages/web3-eth/src/utils/prepare_transaction_for_signing.ts [hunk 3]
```diff
@@ -125,19 +125,19 @@ export const prepareTransactionForSigning = async (
 	web3Context: Web3Context<EthExecutionAPI>,
 	privateKey?: HexString | Uint8Array,
 	fillGasPrice = false,
+	fillGas = true,
```
Comment (jdevcs) [inline packages/web3-eth/src/utils/prepare_transaction_for_signing.ts]:
```suggestion
	fillGasLimit = true,
```

gas and gaslimit are same , it might be confusing with gasPrice, so suggesting to use gaslimit

File: packages/web3-eth/src/types.ts [hunk 4]
```diff
@@ -77,6 +77,7 @@ export interface SendTransactionOptions<ResolveType = TransactionReceipt> {
 	transactionResolver?: (receipt: TransactionReceipt) => ResolveType;
 	contractAbi?: ContractAbi;
 	checkRevertBeforeSending?: boolean;
+	ignoreFillingGas?: boolean;
```
Comment (jdevcs) [inline packages/web3-eth/src/types.ts]:
```suggestion
	ignoreFillingGasLimit?: boolean;
```
is this for gas limit right?

File: packages/web3-eth/src/utils/prepare_transaction_for_signing.ts [hunk 5]
```diff
@@ -125,19 +125,19 @@ export const prepareTransactionForSigning = async (
...
 ) => {
 	const populatedTransaction = (await transactionBuilder({
 		transaction,
 		web3Context,
 		privateKey,
 		fillGasPrice,
+		fillGas,
```
Comment (jdevcs) [inline packages/web3-eth/src/utils/prepare_transaction_for_signing.ts]:
```suggestion
		fillGasLimit,
```

File: packages/web3-eth/src/utils/transaction_builder.ts [hunk 6]
```diff
@@ -129,8 +129,8 @@ export async function defaultTransactionBuilder<ReturnType = Transaction>(option
 	web3Context: Web3Context<EthExecutionAPI & Web3NetAPI>;
 	privateKey?: HexString | Uint8Array;
 	fillGasPrice?: boolean;
+	fillGas?: boolean;
```
Comment (jdevcs) [inline packages/web3-eth/src/utils/transaction_builder.ts]: same here

File: packages/web3-eth/src/utils/transaction_builder.ts [hunk 7]
```diff
@@ -238,7 +236,21 @@ export async function defaultTransactionBuilder<ReturnType = Transaction>(option
...
+	) {
+		const fillGas = await estimateGas(
+			options.web3Context,
+			populatedTransaction,
+			'latest',
+			ETH_DATA_FORMAT,
+		);
+		populatedTransaction = {
+			...populatedTransaction,
+			gas: format({ format: 'uint' }, fillGas as Numbers, ETH_DATA_FORMAT),
```
Comment (jdevcs) [inline packages/web3-eth/src/utils/transaction_builder.ts]: can you point me in 1.x where we are setting gasLimit ( gas ) ?

File: packages/web3-eth/src/utils/transaction_builder.ts [hunk 8]
```diff
@@ -238,7 +236,21 @@ export async function defaultTransactionBuilder<ReturnType = Transaction>(option
 				ETH_DATA_FORMAT,
 			)),
 		};
-
+	if (
+		(isNullish(populatedTransaction.gas) || isNullish(populatedTransaction.gasLimit)) &&
```
Comment (jdevcs) [inline packages/web3-eth/src/utils/transaction_builder.ts]:
```suggestion
		(isNullish(populatedTransaction.gas) && isNullish(populatedTransaction.gasLimit)) &&
```
if both gas and gasLimit are not available then we need to set gas, else one of this should be used to set gas.

File: packages/web3-eth/src/utils/transaction_builder.ts (hunk 7 above)
Comment (luu-alex) [reply packages/web3-eth/src/utils/transaction_builder.ts]: Ah I don't believe we set it in 1.x sorry for the confusion, this is [available](https://github.com/ethers-io/ethers.js/blob/1dd542721d3a9e74d10a92847200f4b2aa5da618/src.ts/providers/abstract-signer.ts#L105) in ethers and would be a solution to solving the transaction evm revert errors [reported](https://github.com/web3/web3.js/issues/6249) by users.

**Comment by luu-alex** [comment]
To avoid breaking changes, added the helpful error message as an innerError. You can see below
Screenshot 2023-07-18 at 1 46 55 PM

**Review by jdevcs — COMMENTED** [review]
some of tests are not passing.

**Comment by jdevcs** [comment]
lgtm,  tests should be fixed. Thanks @luu-alex

**Comment by luu-alex** [comment]
@jdevcs [testcases passing](https://github.com/web3/web3.js/actions/runs/5591976352/jobs/10270844376) approve when you can, thank you

**Review by jdevcs — APPROVED** [review]

**Review by avkos — APPROVED** [review]
